File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummyNode = ListNode(-1)
        cur = dummyNode

        carry = 0

        while l1 or l2:
            sum = carry

            if l1:
                sum = sum + l1.val
            if l2:
                sum = sum + l2.val
            
            newNode = ListNode(sum % 10)
            carry = sum // 10
            cur.next = newNode
            cur = cur.next

            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        
        if carry:
            carryNode = ListNode(carry)
            cur.next = carryNode

        return dummyNode.next

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int sum = 0;
        ListNode* dummyNode = new ListNode(-1);
        ListNode* cur = dummyNode;
        int carry = 0;
        while(l1 || l2) {
            int sum = carry;

            if(l1) sum = sum + l1 -> val;
            if(l2) sum = sum + l2 -> val;

            ListNode* node = new ListNode(sum % 10);
            carry = sum /= 10;
            cur -> next = node;
            cur = cur -> next;

            if(l1) l1 = l1 -> next;
            if(l2) l2 = l2 -> next;
        }

        if(carry) {
            ListNode* newNode = new ListNode(carry);
            cur -> next = newNode;
        }

        return dummyNode -> next;
    }
};